File: mvvmbase/src/main/java/com/bing/mvvmbase/base/page/BaseDataSourceFactory.java
package com.bing.mvvmbase.base.page;

import androidx.lifecycle.MutableLiveData;
import androidx.paging.DataSource;
import io.reactivex.disposables.CompositeDisposable;

public abstract class BaseDataSourceFactory<DS extends BaseDataSource, K, T> extends DataSource.Factory<K, T> {
	protected CompositeDisposable mCompositeDisposable;
	protected MutableLiveData<DS> mSourceLiveData = new MutableLiveData<>();

	public BaseDataSourceFactory(CompositeDisposable compositeDisposable) {
		mCompositeDisposable = compositeDisposable;
	}

	public MutableLiveData<DS> getSourceLiveData() {
		return mSourceLiveData;
	}

	@Override
	public DataSource<K, T> create() {
		DS source = createDataSource();
		mSourceLiveData.postValue(source);
		return source;
	}

	protected abstract DS createDataSource();
}

File: mvvmbase/src/main/java/com/bing/mvvmbase/base/page/BaseDataSourceFactory.kt
package com.bing.mvvmbase.base.page

import androidx.lifecycle.MutableLiveData
import androidx.paging.DataSource
import io.reactivex.disposables.CompositeDisposable

abstract class BaseDataSourceFactory<DS : BaseDataSource<K, T, *>, K, T>(protected var mCompositeDisposable: CompositeDisposable) : DataSource.Factory<K, T>() {
        var sourceLiveData = MutableLiveData<DS>()
                protected set

        override fun create(): DataSource<K, T> {
                val source = createDataSource()
                sourceLiveData.postValue(source)
                return source
        }

        protected abstract fun createDataSource(): DS
}
